### PR TITLE
[Fix] sdpa_decode: fix fp32_dest_acc_en wrong results (DEST overflow) and hang (subblock divisibility)

### LINKED ISSUE #50564: [Bounty $1,500] [sdpa_decode] fp32_dest_acc_en=True silently produces wrong results for most shapes, and hangs at 96 query heads
### Component / Area

`ops / transformer` — `ttnn.transformer.scaled_dot_product_attention_decode`

### Issue Type

Bad Outputs

### Observed

`scaled_dot_product_attention_decode` silently returns wrong results when the padded query-head count is above 32 (`PNHt > 1`) and `fp32_dest_acc_en=True`. Plain causal decode, no attention mask, no paging: with 64 query heads / 8 KV heads / head_dim 128, the worst per-head PCC is **0.359** instead of 0.999. 32 query heads is fine with either setting, and 64 query heads with `fp32_dest_acc_en=False` is fine, so the trigger is the flag together with `PNHt > 1`. It also needs more than one core per head-batch — `max_cores_per_head_batch=1` gives correct results.

With an **odd** `PNHt` (96 query heads) the same configuration **hangs the device** instead of returning wrong data; recovery needs `tt-smi -r`.

`fp32_dest_acc_en` halves DEST (`dst_size = fp32_dest_acc_en ? 4 : 8`, `sdpa_decode_program_factory.cpp:388`), and that value feeds the QK subblock derivation on the next line (`qk_out_subblock_h = min(PNHt, dst_size / qk_out_subblock_w)`) and `MUL_BCAST_GRANULARITY` at `:782`. Both are functions of `PNHt`, so they only change behaviour once `PNHt > 1`. #48748 already records the same class of corruption for Gemma4's `head_dim=512` decode config; this repro shows it is not head_dim-specific.

This combination looks untested: every sdpa decode unit test hardcodes `fp32_dest_acc_en=False`, and none uses more than 32 query heads.

### Expected

Either correct results with `fp32_dest_acc_en=True`, or reject the combination in `validate()`. Silently applying a wrong attention output is the worst of the three outcomes.

### 1. Steps (exact commands)

```bash
cd $TT_METAL_HOME && source python_env/bin/activate
python repro_sdpa_decode_fp32_dest_acc.py          # wrong results
python repro_sdpa_decode_fp32_dest_acc.py hang     # WARNING: hangs the device, needs tt-smi -r
```

<details><summary>repro_sdpa_decode_fp32_dest_acc.py (self-contained)</summary>

```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
#
# SPDX-License-Identifier: Apache-2.0

"""Repro: sdpa_decode returns silently wrong results with fp32_dest_acc_en=True
when the padded query-head count exceeds 32 (PNHt > 1)."""
import math
import sys

import torch

import ttnn

HD, B, S, NKV, CUR_POS = 128, 1, 256, 8, 200


def pcc(a, b):
    a, b = a.flatten().double(), b.flatten().double()
    va, vb = a - a.mean(), b - b.mean()
    return float((va @ vb) / (va.norm() * vb.norm()))


def run(device, num_q_heads, fp32_dest_acc, max_cores=16):
    torch.manual_seed(0)
    k = torch.randn(B, NKV, S, HD).bfloat16().float()
    v = torch.randn(B, NKV, S, HD).bfloat16().float()
    q = torch.randn(1, B, num_q_heads, HD).bfloat16().float()
    cur_pos = torch.tensor([CUR_POS] * B, dtype=torch.int32)
    scale = 1.0 / math.sqrt(HD)

    tt = lambda t, dt: ttnn.Tensor(t, dt).to(ttnn.TILE_LAYOUT).to(device)
    out = ttnn.transformer.scaled_dot_product_attention_decode(
        tt(q, ttnn.bfloat16),
        tt(k, ttnn.bfloat16),
        tt(v, ttnn.bfloat16),
        cur_pos_tensor=ttnn.Tensor(cur_pos, ttnn.int32).to(device),
        is_causal=True,
        scale=scale,
        program_config=ttnn.SDPAProgramConfig(
            compute_with_storage_grid_size=device.compute_with_storage_grid_size(),
            q_chunk_size=32,
            k_chunk_size=64,
            exp_approx_mode=False,
            max_cores_per_head_batch=max_cores,
        ),
        compute_kernel_config=ttnn.init_device_compute_kernel_config(
            device.arch(),
            math_fidelity=ttnn.MathFidelity.HiFi2,
            math_approx_mode=True,
            fp32_dest_acc_en=fp32_dest_acc,
            packer_l1_acc=False,
        ),
    )
    got = ttnn.to_torch(out)[0, 0, :num_q_heads, :].float()

    group = num_q_heads // NKV
    worst = min(
        pcc(
            torch.softmax((k[0, i // group, : CUR_POS + 1] @ q[0, 0, i]) * scale, dim=-1)
            @ v[0, i // group, : CUR_POS + 1],
            got[i],
        )
        for i in range(num_q_heads)
    )
    return worst


device = ttnn.open_device(device_id=0)
try:
    for nh in (32, 64):
        for fp32 in (False, True):
            worst = run(device, nh, fp32)
            print(
                f"q_heads={nh:3d} (PNHt={nh // 32})  fp32_dest_acc_en={str(fp32):5s}  "
                f"worst per-head PCC {worst:.6f}  {'ok' if worst > 0.99 else '<-- WRONG'}",
                flush=True,
            )
    if len(sys.argv) > 1 and sys.argv[1] == "hang":
        print("96 q heads (PNHt=3) + fp32_dest_acc_en=True -> expect device hang", flush=True)
        run(device, 96, True)
        print("did not hang", flush=True)
finally:
    ttnn.close_device(device)
```

</details>

### 2. Input data / link or description

None — the script generates its own seeded random Q/K/V.

### 3. Frequency

Always (deterministic, both the wrong results and the hang).

### System Details

**1. Software Versions**

```
OS: Ubuntu 24.04.3 LTS (kernel 6.17.0-35-generic)
tt-metal: 0d4af3ae6e2c9aa226aebbe84350a6e0540a24e6 (main, 2026-07-17)
Driver: TT-KMD 2.7.0
Firmware bundle: 19.7.0.0
```

**2. Hardware Details**

```
Product: Blackhole
Card/System: p300c (repro runs on a single chip, device_id=0)
```

### Is this a regression?

No

### Regression Details

```
First bad version:   n/a - this configuration has never worked
Last known good:     none
Git bisect status:   not bisected; established from history (see below)
```

More than 32 query heads only started producing correct results at all with #34113 (68459b0e25, 2025-12-10), which fixed #33440. The three expressions that make the fp32 path wrong are unchanged since then — `dst_size = fp32_dest_acc_en ? 4 : 8`, `qk_out_subblock_h = min(PNHt, dst_size / qk_out_subblock_w)` (and the `out_out_subblock_h` twin), and `MUL_BCAST_GRANULARITY = min(PNHt * Sk_chunk_t, dst_size)`. They date back to 1901aedf96 (2024-08-19, the move of SDPA to ttnn); #36440 only moved the granularity into a helper without changing the formula. So `fp32_dest_acc_en=True` with `PNHt > 1` has never worked, rather than having regressed.

### Logs & Diagnostics

```
q_heads= 32 (PNHt=1)  fp32_dest_acc_en=False  worst per-head PCC 0.999664  ok
q_heads= 32 (PNHt=1)  fp32_dest_acc_en=True   worst per-head PCC 0.999605  ok
q_heads= 64 (PNHt=2)  fp32_dest_acc_en=False  worst per-head PCC 0.999362  ok
q_heads= 64 (PNHt=2)  fp32_dest_acc_en=True   worst per-head PCC 0.359092  <-- WRONG
96 q heads (PNHt=3) + fp32_dest_acc_en=True -> expect device hang
<hangs here; no further output, device needs tt-smi -r>
```

### PR DESCRIPTION
**Pull request by kinginu**
### PR Category

Bug fix — closes https://github.com/tenstorrent/tt-metal/issues/50564. Two independent `fp32_dest_acc_en=True` defects on `scaled_dot_product_attention_decode`, both because fp32 accumulation halves DEST from 8 to 4 tiles.

### 1. Wrong results — `correction_block` overflows DEST

The cross-core flash-decode softmax correction runs five tiles through DEST in one `tile_regs_acquire` (prev_max, worker_max, cur_max, prev_sum, worker_sum). That fits in 8 tiles but not 4, so `dst_reg_4` is out of range and the worker sum is garbage — attention weights summed to 4–47 instead of 1. Split into two ≤3-tile passes when DEST cannot hold five, gated on `compute_kernel_lib::DEST_AUTO_LIMIT`; the bf16 path keeps the fused kernel, bit-identical.

### 2. Device hang — matmul subblock height must divide PNHt

The QK / out matmuls set `subblock_h = min(PNHt, dst_size / subblock_w)`; `matmul_blocks` loops `PNHt / subblock_h` times over `subblock_h` rows. When `subblock_h` does not divide `PNHt` the integer division truncates, the CB is under-produced, and the consumer waits forever. fp32 caps the subblock at 2, so an odd `PNHt` (e.g. 3 = 96 query heads, k_chunk 64) covers only 2 of 3 rows and hangs. Clamp the cap to the largest divisor of `PNHt`.

### Verification

Worst per-head PCC vs torch, plain causal decode, 8 KV heads, on Blackhole p150. `fp32_dest_acc_en=False` unchanged throughout.

- Wrong results: 12 fp32 cells over head_dim × heads × k_chunk go from 0.010–0.481 to **0.9995+**.
- Hang: 96 (PNHt=3) and 160 (PNHt=5) query heads at k_chunk 64 hang on main and now run at **0.9995**.

Tests: `test_sdpa_decode_fp32_dest_acc` (wrong results) and `test_sdpa_decode_fp32_dest_acc_odd_pnht` (hang). Both fail on main — the first at PCC 0.086, the second by hanging — and pass here. `sdpa_test_utils` gains an `fp32_dest_acc_en` argument defaulting to `False`. Local runs: `test_sdpa_decode.py` 14 passed / 1 skipped, `test_sdpa_prefill.py` 8 passed / 2 skipped.

Re-verified after merging main, on Blackhole P300 and on Wormhole n300. The head_dim x q_heads x k_chunk sweep is correct in all four (`fp32_dest_acc_en`, `dst_full_sync_en`) combinations, PNHt=3 and PNHt=5 included, and `test_sdpa_decode.py` plus `test_sdpa_prefill.py` give 22 passed / 3 skipped on both.

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp [hunk 1]
```diff
@@ -737,26 +738,78 @@ void correction_block(
...
+    } else {
+        for (uint32_t i = 0; i < num_head_tiles; i++) {
+            tile_regs_acquire();
+            copy_tile_to_dst_init_short(cb_worker_max);
+            exp_tile_init<EXP_APPROX_MODE>();
+            copy_tile(cb_prev_max, i, dst_reg_0);
+            copy_tile(cb_worker_max, i, dst_reg_1);
+            copy_tile(cb_prev_sum, i, dst_reg_3);
+            copy_tile(cb_worker_sum, i, dst_reg_4);
+            MATH((fused_max_sub_exp_add_tile<vector_mode>(0, scale_bf16)));
```
Comment (mateusznowakTT) [inline ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp:817]: I believe the fix could be achieved either by improving fused_max_sub_exp_add_tile or making sure it is configured properly in program factory without resolving to 2 passes. Could you please check this, especially the latter?

**Review by mateusznowakTT — CHANGES_REQUESTED** [review]

**Comment by kinginu** [comment]
Tried both routes. Neither beats the split.

Program factory first. The only lever there is `dst_full_sync_en`, since that is what decides how much DEST the fp32 path gets:

| | `dst_full_sync_en=false` | `dst_full_sync_en=true` |
|---|---|---|
| `fp32_dest_acc_en=false` | 8 | 16 |
| `fp32_dest_acc_en=true` | **4** | **8** |

So configuring the split away means forcing full sync whenever fp32 is set, which trades a problem local to the cross core reduction for losing math/pack overlap everywhere. It does work and every shape is correct, but Blackhole P300, 100 iterations per cell:

| head_dim / q_heads / k_chunk / seq | base | this PR | full sync | vs this PR |
|---|---|---|---|---|
| 128 / 32 / 64 / 8192 | 121.0us | 121.7us | 122.3us | +0.5% |
| 128 / 64 / 64 / 8192 | 137.8us | 137.2us | 142.8us | **+4.1%** |
| 128 / 32 / 128 / 32768 | 381.5us | 381.9us | 383.2us | +0.3% |
| 128 / 64 / 128 / 32768 | 395.0us | 393.4us | 399.6us | **+1.6%** |
| 256 / 32 / 64 / 8192 | 219.2us | 220.1us | 221.8us | +0.8% |
| 512 / 32 / 64 / 8192 | 402.7us | 402.9us | 409.0us | **+1.5%** |
| 512 / 32 / 128 / 8192 | 426.4us | 426.1us | 431.8us | +1.3% |
| 128 / 32 / 512 / 32768 | 443.3us | 444.7us | 445.8us | +0.2% |
| 128 / 32 / 1024 / 32768 | 539.7us | 540.1us | 540.4us | +0.1% |

base and this PR are within noise of each other in every cell, so the split costs nothing measurable, it only runs in the reduction. Forcing full sync is slower in all nine, worst at PNHt=2 where the reduction runs most often. Giving full sync to the reducer cores only, by emitting the compute kernel twice, is worse again: a flat 35us per invocation for the extra kernel.

`fused_max_sub_exp_add_tile` can be brought down to four tiles, since `worker_sum` is dead once its product is formed and `cur_max` can take that slot. The obvious version does not compile (sfpi register spill), loading the sums only after the maxes are dead does. Correct on the whole sweep, and a wash on perf:

| head_dim / q_heads / k_chunk / seq | this PR | four tile fused | vs this PR |
|---|---|---|---|
| 128 / 32 / 64 / 8192 | 121.7us | 121.8us | +0.1% |
| 128 / 64 / 64 / 8192 | 137.2us | 139.5us | +1.7% |
| 128 / 32 / 128 / 32768 | 382.2us | 382.4us | +0.1% |
| 128 / 64 / 128 / 32768 | 393.7us | 396.8us | +0.8% |
| 256 / 32 / 64 / 8192 | 220.3us | 220.3us | 0.0% |
| 512 / 32 / 64 / 8192 | 403.8us | 404.0us | 0.0% |
| 512 / 32 / 128 / 8192 | 426.6us | 426.5us | 0.0% |
| 128 / 32 / 512 / 32768 | 443.8us | 444.3us | +0.1% |
| 128 / 32 / 1024 / 32768 | 539.8us | 539.2us | -0.1% |

Branch is [here](https://github.com/kinginu/tt-metal/compare/fix/sdpa-decode-fp32-dest-correction...fix/sdpa-fused-4dst) if you want it, one commit on top of this PR, +92 / -140, same test results. I would rather not carry it in this PR since it pulls in the SFPU headers for both architectures and I can only measure Blackhole.

Your point about the gate was right though: it keyed on `DST_ACCUM_MODE`, which is only half of what sets DEST capacity. It now gates on `compute_kernel_lib::DEST_AUTO_LIMIT < 5`, so fp32 together with `dst_full_sync_en` keeps the fused kernel. Pushed.

`test_sdpa_decode.py` 14 passed / 1 skipped, `test_sdpa_prefill.py` 8 passed / 2 skipped, repro from #50564 clean.

File: ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp (hunk 1 above)
Comment (kinginu) [reply ttnn/cpp/ttnn/operations/transformer/sdpa/device/kernels/compute/compute_common.hpp:817]: Answered in the comment below: forcing dst_full_sync_en is correct but up to 4.1% slower, the reducer only variant is worse, and the four tile LLK version is a wash (branch linked). The gate you pointed at is fixed, it keys on DEST_AUTO_LIMIT < 5 now. Let me know if you'd rather have the four tile version and I'll swap it in.

**Comment by kinginu** [comment]
@jberkowitzTT could you run /codeowners ping?

**Review by yieldthought — APPROVED** [review]

**Comment by kinginu** [comment]
@mateusznowakTT could you take another look?

**Review by mateusznowakTT — APPROVED** [review]
